### ISSUE TITLE
Testing features

### ISSUE THREAD

**Issue by NickSeagull** [neohaskell/NeoHaskell#33]
- test: Run tests
- test-watch: Run tests in watch mode, rerunning on file changes

**Comment by moodmosaic** [comment]
>test: Run tests

With this command we may also be able to support property-based tests.

### Discoverability

We should be able to execute as a test any function that meets the following criteria:

* The function is public.
* The function name starts with `test_` (configurable).
* If a public function  named `beforeEach` is present it can be executed before each test is run.

<table>
  <thead>
    <tr>
      <th>type</th>
      <th>prefix</th>
      <th>arguments</th>
      <th>semantics</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>concrete</td>
      <td><code class="language-plaintext highlighter-rouge">test_</code></td>
      <td>no</td>
      <td>single execution with concrete values</td>
    </tr>
    <tr>
      <td>property</td>
      <td><code class="language-plaintext highlighter-rouge">test_</code></td>
      <td>yes</td>
      <td>multiple executions with randomly generated concrete values (smaller values)</td>
    </tr>
    <tr>
      <td>fuzz</td>
      <td><code class="language-plaintext highlighter-rouge">test_</code></td>
      <td>yes</td>
      <td>multiple executions with randomly generated concrete values (<a href="https://github.com/dubzzz/fast-check/discussions/2978">biased</a>)<sup>*</sup></td>
    </tr>
  </tbody>
</table>

<sup>*</sup>Also configurable.

---

For the biased case, depending on the underlying testing library (QuickCheck, hedgehog, falsify), there may exist a command that overrides the scope and shrinking of generated values so the tests can run in 'fuzz' mode. For example:

* `test`: Run tests
* `fuzz`: Run tests (biased)

**Comment by NickSeagull** [comment]
This looks awesome!

One of the things that I was thinking to promote was the usage of doctests. Many times folks skip writing tests because it is an additional task to do, and an afterthought in some way.

My hypothesis is that if we teach folks to describe and document their functions, as well as giving an example, we start getting a bit of test coverage, which is better than nothing.

On top of that, now with GitHub Copilot, writing a good documentation, along with examples (doc tests) the implementation of any code becomes much easier.

[They also support property based testing](https://github.com/sol/doctest#writing-examples-and-properties), which is cool

**Comment by moodmosaic** [comment]
Yes! I have been [using doctest even in F#](https://github.com/moodmosaic/doctest) in the past and I agree with your thoughts around it.
